resource.go

```diff
@@ -60,13 +82,79 @@ func getResourceField(v interface{}) (f *structs.Field, err error) {
 	return
 }
 
+func getIdFromMap(m map[string]interface{}) (ret int, err error) {
+	if val, ok := m["id"].(int); !ok {
+		ret = 0
+	} else {
+		ret = val
+	}
+	return
+}
+
 // Retrieve Id field from interface.
 func resourceId(v interface{}) (ret int, err error) {
 	var f *structs.Field
+	invalidTypeErr := errors.New("Not a struct, map, or pointer to either")
+	ty := reflect.TypeOf(v)
 
-	if f, err = getResourceField(v); err == nil {
-		ret = f.Value().(int)
+	switch ty.Kind() {
+	case reflect.Map:
+		if m, ok := v.(map[string]interface{}); !ok {
+			err = errors.New("Invalid map")
+		} else {
+			return getIdFromMap(m)
+		}
+	case reflect.Struct:
+		if f, err = getResourceField(v); err == nil {
+			ret = f.Value().(int)
+		}
+	case reflect.Ptr:
+		return resourceId(reflect.Indirect(reflect.ValueOf(v)).Interface())
+	default:
+		err = invalidTypeErr
 	}
 
 	return
 }
+
+// Determine whether the struct's id will be omitted in json encoding.
+func isIdOmitted(v interface{}) (bool, error) {
+
+	if structs.IsStruct(v) {
```

Comment:
I always think `if ... { <big chunk of code> } else { <other big chunk of code> }` is an indicator of the need for a refactor - it's harder to read and this seems to me to indicate that each block of code is doing something distinct which could be nicely self-documented with a separate function.

Perhaps moving to `isStructIdOmitted()` and `isMapIdOmitted()` would work better?